samples/hello-cube.html

```diff
@@ -0,0 +1,336 @@
+}
+
+fragment float4 fragment_main(FragmentData data [[stage_in]])
+{
+    return float4(data.color);
+}
+`
+
+// A * indicates that this symbol has yet to be implemented in WebKit.
+
+let device, context, verticesBuffer, mappedTransformArray, uniformsBindGroup, pipeline, renderPassDescriptor, queue;
+let projectionMatrix = mat4.create();
+
+const vertexSize = 4 * 8;
+const verticesSize = vertexSize * 36;
+const colorOffset = 4 * 4;
+function populateVerticesArray(array) {
+    const vertexData = [
+        // float4 position, float4 color
+        1, -1, 1, 1, 1, 0, 1, 1,
+        -1, -1, 1, 1, 0, 0, 1, 1,
+        -1, -1, -1, 1, 0, 0, 0, 1,
+        1, -1, -1, 1, 1, 0, 0, 1,
+        1, -1, 1, 1, 1, 0, 1, 1,
+        -1, -1, -1, 1, 0, 0, 0, 1,
+
+        1, 1, 1, 1, 1, 1, 1, 1,
+        1, -1, 1, 1, 1, 0, 1, 1,
+        1, -1, -1, 1, 1, 0, 0, 1,
+        1, 1, -1, 1, 1, 1, 0, 1,
+        1, 1, 1, 1, 1, 1, 1, 1,
+        1, -1, -1, 1, 1, 0, 0, 1,
+
+        -1, 1, 1, 1, 0, 1, 1, 1,
+        1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1, -1, 1, 1, 1, 0, 1,
+        -1, 1, -1, 1, 0, 1, 0, 1,
+        -1, 1, 1, 1, 0, 1, 1, 1,
+        1, 1, -1, 1, 1, 1, 0, 1,
+
+        -1, -1, 1, 1, 0, 0, 1, 1,
+        -1, 1, 1, 1, 0, 1, 1, 1,
+        -1, 1, -1, 1, 0, 1, 0, 1,
+        -1, -1, -1, 1, 0, 0, 0, 1,
+        -1, -1, 1, 1, 0, 0, 1, 1,
+        -1, 1, -1, 1, 0, 1, 0, 1,
+
+        1, 1, 1, 1, 1, 1, 1, 1,
+        -1, 1, 1, 1, 0, 1, 1, 1,
+        -1, -1, 1, 1, 0, 0, 1, 1,
+        -1, -1, 1, 1, 0, 0, 1, 1,
+        1, -1, 1, 1, 1, 0, 1, 1,
+        1, 1, 1, 1, 1, 1, 1, 1,
+
+        1, -1, -1, 1, 1, 0, 0, 1,
+        -1, -1, -1, 1, 0, 0, 0, 1,
+        -1, 1, -1, 1, 0, 1, 0, 1,
+        1, 1, -1, 1, 1, 1, 0, 1,
+        1, -1, -1, 1, 1, 0, 0, 1,
+        -1, 1, -1, 1, 0, 1, 0, 1,
+    ];
+
+    for (let i = 0; i < vertexData.length; ++i) {
+        array[i] = vertexData[i];
+    }
+}
+
+const transformSize = 4 * 16;
+function populateTransformArray(array) {
+    let viewMatrix = mat4.create();
+    mat4.translate(viewMatrix, viewMatrix, vec3.fromValues(0, 0, -5));
+    let now = Date.now() / 1000;
+    mat4.rotate(viewMatrix, viewMatrix, 1, vec3.fromValues(Math.sin(now), Math.cos(now), 0));
+
+    let modelViewProjectionMatrix = mat4.create();
+    mat4.multiply(modelViewProjectionMatrix, projectionMatrix, viewMatrix);
+
+    for (let i = 0; i < 16; i++) {
+        array[i] = modelViewProjectionMatrix[i];
+    }
+}
+
+async function init() {
+    const adapter = await webgpu.requestAdapter({}); // Current (unifinished) implementation of requestAdapter requires this argument.
+    device = adapter.createDevice();
+
+    const canvas = document.querySelector('canvas');
+
+    const aspect = Math.abs(canvas.width / canvas.height);
+    mat4.perspective(projectionMatrix, (2 * Math.PI) / 5, aspect, 1, 100.0);
+
+    context = canvas.getContext('webgpu');
+
+    const swapChainDescriptor = { 
+        device: device, 
+        format:"B8G8R8A8Unorm", 
```

Comment:
This is sort of pending on figuring exactly how swapchain formats work, but we agreed in the meeting that there would be 2 always-supported formats (probably RGBA8 or BGRA8, and F16) and optionally some platform-specific "optimal" formats. This is needed for optimal interactions with the system's compositor/overlays.

Reply:
It is more natural that the guaranteed 8bit format be RGBA8 but we could also guarantee BGRA8 because that's the optimal format on at least macOS and Windows.

Reply:
Agreed, BGRA8 might be a better always-supported option for that reason.

Reply:
It's BGRA8 because that's the only format I've tested. In addition, this field is currently ignored by the WebKit impl and I actually have it hard coded to BGRA8.